library-best-practices/JLBP-1.md

```diff
@@ -0,0 +1,23 @@
+[JLBP-1] Minimize dependencies
+------------------------------
+
+Use the minimum number of dependencies that is reasonable.
+Adding a dependency for a large amount of functionality may be ok,
+but avoid pulling in dependencies just to save a few lines of code,
+because every dependency is a liability.
+
+Some specific notes about minimizing dependencies:
+
+- Use the smallest scope possible. For example, auto-value doesn't
+  need to use compile scope, and can instead use compile-only,
+  since it doesn't need to appear on the classpath of consumers.
+  - All libraries used only for testing should have test scope
+    (e.g. junit, mockito, and truth).
+- Scrutinize all dependency additions. Check the result of
+  `mvn dependency:tree` (after running `mvn install -DskipTests`
+  to build the library) to see which transitive dependencies are
+  added by just adding a single dependency to your own library,
+  and if you require all of the transitive dependencies.
+- Avoid cyclic dependencies
+  - E.g., gRPC depends on OpenCensus which depends on gRPC.
+  - Maven can detect this.
```

Comment:
It's a little stronger: Maven does not allow this. Build fails outright if there's a cyclic dependency.

Reply:
@saturnism Looking at grpc's dependencies, there is no cycle technically. It goes `grpc-core` -> `opencensus-api` -> `grpc-context`. Maybe you're talking about an inter-project cycle, which is actually not detectable by maven?

Reply:
Thinking about this more, I think it should be a separate item - it's more subtle than just minimizing dependencies. I have pulled it offline (see internal doc).